### ISSUE TITLE
Add the ability to fetch tiles in parallel

### ISSUE THREAD

**Issue by tcihak-fqa** [geopandas/contextily#215]
Fetching the web tiles can take a long time when there are many to download.
Would it be possible to do multiple tile requests in parallel using the multiprocessing module?

**Comment by tcihak-fqa** [comment]
Thanks for the response! So I'm using contextily to generate static maps from a web api that has high bandwidth. I agree that most of the time is spent on I/O latency.

**Comment by JacobJeppesen** [comment]
I have actually been considering making a pull request for this exact functionality. I've been playing around with it in another project, and for large files it can make a rather big difference. The optimal number of parallel downloads differ quite a bit from endpoint to endpoint, but 8-16 is normally a good range. The code below can be used as a drop-in replacement for the for loop:

```
def bounds2img(
        w, s, e, n, zoom="auto", source=None, ll=False, wait=0, max_retries=2, num_parallel_tile_downloads=8
):
.
.
.
    # download and merge tiles
    # tiles = []
    # arrays = []
    # for t in mt.tiles(w, s, e, n, [zoom]):
    #     x, y, z = t.x, t.y, t.z
    #     tile_url = provider.build_url(x=x, y=y, z=z)
    #     image = _fetch_tile(tile_url, wait, max_retries)
    #     tiles.append(t)
    #     arrays.append(image)
    from joblib import Parallel, delayed  # This should go to the top of the file
    tiles = list(mt.tiles(w, s, e, n, [zoom]))
    tile_urls = [provider.build_url(x=tile.x, y=tile.y, z=tile.z) for tile in tiles]
    max_num_parallel_tile_downloads = 32
    # Note: num_parallel_tile_downloads is currently an argument in the function
    if num_parallel_tile_downloads < 1 or num_parallel_tile_downloads > max_num_parallel_tile_downloads:
        raise ValueError(
            f"num_parallel_tile_downloads must be between 1 and {max_num_parallel_tile_downloads}"
        )
    arrays = \
        Parallel(n_jobs=num_parallel_tile_downloads, prefer="threads")(
            delayed(_fetch_tile)(tile_url, wait, max_retries) for tile_url in tile_urls)

    merged, extent = _merge_tiles(tiles, arrays)
.
.
.
```

I just tested it in the `intro_guide.ipynb` notebook by downloading the ghent image in the *Coordinate-based searches* section, with the following code:
```
west, south, east, north = (
    3.616218566894531,
    50.98912458110244,
    5.8483047485351562,
    54.13994019806845
             )
import time
start_time = time.time()
ghent_img, ghent_ext = cx.bounds2img(west, 
                                     south, 
                                     east, 
                                     north, 
                                     ll=True, 
                                     zoom=10,
                                     source=cx.providers.Stamen.Toner,
                                     num_parallel_tile_downloads=8
                                    )
print(f"Download time: {time.time() - start_time}")
ghent_img.shape 
```

Note that I had to out-comment the `@memory.cache` decorator for the `_fetch_tile` function during the test.

The shape of the image was `(7680, 3584, 4)`, and the download times were:
num_parallel_tile_downloads=1:   69.49s
num_parallel_tile_downloads=2:   36.11s 
num_parallel_tile_downloads=4:   18.32s 
num_parallel_tile_downloads=8:     9.55s 
num_parallel_tile_downloads=16:   5.02s 
num_parallel_tile_downloads=32:   2.92s 

I don't think there should be any downside to always do it in parallel (the overhead should be minimal), as long as the number of parallel downloads don't bomb the endpoint. A default value of 16 is a good starting point in my experience, with the different endpoints I've tested. That normally gives almost linear improvements. Above that, it differs quite a bit. 

That ended up being quite a bit of text. Hopefully it's useful :wink: 

Let me know if you would like a pull request with the change :+1:

**Comment by JacobJeppesen** [comment]
Just made a couple more tests with smaller images:

At `zoom=9` the shape was `(2048, 1024, 4)` and download times were:
num_parallel_tile_downloads=1: 5.38s
num_parallel_tile_downloads=16: 0.50s

At `zoom=7` the shape was `(512, 512, 4)` and download times were:
num_parallel_tile_downloads=1: 0.69s
num_parallel_tile_downloads=16: 0.17s (there's only 4 tiles so in practice we only do 4 parallel downloads)

**Comment by tcihak-fqa** [comment]
This solution looks very nice Jacob! My vote would be to create a PR and get it into a release ... if that counts.
If not, what are the options to use it from my app? I have a devops pipeline that builds a docker image containing the app.
I don't experience with creating forks or applying patches.

**Comment by tcihak-fqa** [comment]
Thanks Jacob.
I monkey patched your original solution and it seems to be working well. I haven't encountered any memory issues but the number of api requests has been light so far.

**Comment by martinfleis** [comment]
Closed by #217
